File: backend/app/api/schemas/policy.py
"""Read-only policy source registry response."""

from datetime import date, datetime

from app.domain import PolicySource, PolicySourceStatus

from .common import ApiModel, DemoMetadata


class PolicySourceResponse(ApiModel):
    source_id: str
    authority: str
    title: str
    document_type: str
    published_at: date | None
    effective_from: date | None
    effective_to: date | None
    reference_url: str | None
    corroborating_urls: tuple[str, ...]
    verified_at: datetime | None
    scope: str | None
    notes: str | None
    status: PolicySourceStatus
    demo: DemoMetadata = DemoMetadata()

    @classmethod
    def from_source(cls, source: PolicySource) -> "PolicySourceResponse":
        return cls(
            source_id=source.source_id,
            authority=source.authority,
            title=source.title,
            document_type=source.document_type,
            published_at=source.published_at,
            effective_from=source.effective_from,
            effective_to=source.effective_to,
            reference_url=(
                str(source.reference_url) if source.reference_url else None
            ),
            corroborating_urls=tuple(
                str(url) for url in source.corroborating_urls
            ),
            verified_at=source.verified_at,
            scope=source.scope,
            notes=source.notes,
            status=source.status,
        )

File: frontend/src/lib/api/policy.ts
import { apiRequest, ClaimSaathiApiError } from "./client";
import { hasSafeDemoMetadata } from "./contracts";
import type { PolicySourceResponse } from "./types";

function assertPolicySource(value: PolicySourceResponse): PolicySourceResponse {
  const nullableStrings = [
    value?.published_at,
    value?.effective_from,
    value?.effective_to,
    value?.reference_url,
    value?.verified_at,
    value?.scope,
    value?.notes,
  ];
  if (
    !value ||
    typeof value.source_id !== "string" ||
    typeof value.authority !== "string" ||
    typeof value.title !== "string" ||
    typeof value.document_type !== "string" ||
    !nullableStrings.every(
      (item) => item === null || typeof item === "string",
    ) ||
    !Array.isArray(value.corroborating_urls) ||
    !value.corroborating_urls.every((item) => typeof item === "string") ||
    !["ACTIVE", "INACTIVE", "SUPERSEDED", "REVIEW_REQUIRED"].includes(
      value.status,
    ) ||
    !hasSafeDemoMetadata(value.demo)
  ) {
    throw new ClaimSaathiApiError(
      "INVALID_POLICY_SOURCE_RESPONSE",
      "The reviewed source could not be displayed safely.",
      200,
    );
  }
  return value;
}

export async function getPolicySource(
  sourceId: string,
): Promise<PolicySourceResponse> {
  const response = await apiRequest<PolicySourceResponse>(
    `/api/v1/policy/sources/${encodeURIComponent(sourceId)}`,
  );
  return assertPolicySource(response);
}
